### PR TITLE
Blocking(melee)

### PR DESCRIPTION
**Pull request by ErisMeleeEnjoyer**
## About The Pull Request
blocked
Remove passive melee blocking that scaled off of ROB 
Add active blocking - a verb for humans that allows them to block an incoming attack once. It has no limits on how long it can last, but you are slowed down while you block. Blocking is turned off when you attack someone or get grabbed.

Since it's a verb, it can be macro'd. A button on hud has been added as well.
Visual indicator of blocking is added.
Hotkey for blocking - F. Old "cycle-intent" hotkeys have been removed due to me lacking any space for new hotkeys(G is already reserved for another feature in the near future).

image
image
image
image

Blocking reduces damage taken from a melee hit; your TGH stat affects how much damage you can block.
If you block with your bare hands, you can block up to 0.3 of your TGH stat, if you block with an item in your hands, you block up to 0.2 of your TGH stat PLUS the item's size multiplied by five.
image

## Why It's Good For The Game
Remove redundant RNG from PVP.
In practice, it's very hard to get enough toughness(200) to completely negate _MELEE_ attacks. A brief reminder, to block a 30 damage attack, a regular unmodded sword, you would need to have 120 TGH. To block a 50 damage attack, like a good toolmodded sword, you would need around twice as much stats. It'd be more viable to max our robustness for those sweet instant grabs(gonna be nerfed next week).
Shields that have passive block chance depending on robustness are not affected. This active blocking is inferior to shields by design.
Spices up melee combat, makes it more interesting and a bit more complicated than clicking until horizontal. Will go well with melee rework from kotmap in the nearest future.
As for making grabs too strong, a nerf for those will be coming next week, and it won't be some number change or a cap.

## Changelog
:cl: Kegdo
add: blocking
add: blocking button on HUD
add: block hotkey(f)
add: block indicator
remove: old cycle-intent hotkeys
/:cl:

### CODE REVIEW

**Comment by ErisMeleeEnjoyer** [comment]
lemme figure out an issue with the HUD i've found, and then it's going to be completely done.

**Comment by assortedbeads** [comment]
When the IH commander tried to punch you out but you've been toughmaxxing all shift

[](https://youtu.be/WHujzgKlPqM)

**Comment by SirRichardFrancis** [comment]
Another circle of balancing hell? Can we not have that?

**Comment by GrayRachnid** [comment]
I recommend looking at CitadelStation's blocking and taking inspiration from it. They've had it for 3 years now and it's not overpowering. Completely negating attacks is too strong. At most, this should reduce the attack's damage.

**Comment by ErisMeleeEnjoyer** [comment]
> I recommend looking at CitadelStation's blocking and taking inspiration from it. They've had it for 3 years now and it's not overpowering.

I... I am not even checking a build that focuses on furries and has ERP in it.

> Completely negating attacks is too strong. At most, this should reduce the attack's damage.

It reduces attack damage. To completely negate an attack, you need to basically toughmaxx all shift. What's stopping you from maxing out any other stat and becoming a god with that instead?

**Comment by ErisMeleeEnjoyer** [comment]
> I remember very well through how many circles of hell Beestation parrying went before being deleted for being unintuitive mechanic and now I know through how many circles of hell Cit blocking went through. Because of that I really don't fucking like the idea of melee blocking.

Blocking != parrying. For me, it's intuitive that blocking is affected by toughness, but we can also make an article on wiki about this(hello grab-fu page) and add a tip about that in game.
I PRESUME that these servers had troubles with parrying and blocking since they did not have stats/combat traits, unlike us. Since, you know, those servers are ~~bad~~ under the rule that not everyone is able to disarm someone else because of them not having skills or some other ridicilous idiocy like that.

> Besides, this also needs a hotkey. Don't you even think about bringing up holster hotkey rant.

Hotkey - MMB. Since this is also a verb, anyone can macro it for their own hotkey if they don't like MMB.

**Comment by SirRichardFrancis** [comment]
Know what else MMB is? Swapping hands. And using RIG module.

**Comment by ErisMeleeEnjoyer** [comment]
yes, and no one(recorded in history) uses swapping hands on MMB since hotkeys(X) exist, and the SWAP button above them exists as well. RIG modules overwrite MMB hotkey, and ONLY if someone sets their RIG module key to MMB; which means, if they need to use a module and block, they can either macro blocking to something else, or change module key.

**Comment by ValoTheValo** [comment]
> yes, and no one(recorded in history) uses swapping hands on MMB since hotkeys(X) exist, and the SWAP button above them exists as well. RIG modules overwrite MMB hotkey, and ONLY if someone sets their RIG module key to MMB; which means, if they need to use a module and block, they can either macro blocking to something else, or change module key.

i use swapping hands on MMB relatively often

**Comment by Firefox13** [comment]
Do not add multiple uses for one hotkey.

**Comment by Firefox13** [comment]
Better yet, add multiple hotkeys for a single thing.

I cannot use MMB because my fucking mousewheel requires me to break my finger in order to execute enough pressure to press it

**Comment by DiaFRAME444** [comment]
TGH is a great stat to have at high already, this basically makes TGH the best stat to max since scaling means at +100 TGH you don't go into paincrit AND can barehand block pretty much any melee attack.

**Comment by ErisMeleeEnjoyer** [comment]
Values adjusted, hotkey changed to F, HUD issue fixed. Ready for merge.

**Comment by GrayRachnid** [comment]
> > I recommend looking at CitadelStation's blocking and taking inspiration from it. They've had it for 3 years now and it's not overpowering.
> 
> I... I am not even checking a build that focuses on furries and has ERP in it.

Why not? They've done a really good job of balancing blocking. Also, while Citadel has Furries and ERP in it they still have competent coders that have created features. Their combat is touted as one of the most unique in SS13 following after LifeWeb.

If you can't take inspiration from a server just because it has 'furries' and 'erp' when you don't even have to interface with it, what are you doing even. Literally just look at the blocking code.
https://github.com/Citadel-Station-13/Citadel-Station-13/blob/0bfb671f4e4239818120b69768546849dfc990ba/code/modules/mob/living/living_active_block.dm

**Comment by GrayRachnid** [comment]
Also if we add blocking without any visual indicators it's kind of shit.

**Comment by GrayRachnid** [comment]
This whole PR stinks of "I did not think through the consequences of this feature and how it would effect the game balance."

**Comment by githubuser4141** [comment]
> Their combat is touted as one of the most unique in SS13 following after LifeWeb.

🦑 
Is it.. really? I've played TG Cit before and it looked like regular TG with some minor additions. Other codebases like Bay, CM, Burgerstation & McDonalds CM feel like they had more unique combat.

I'm not sure when's the last time you played ERIS but the "anti ERP" brigade is pretty common here? Like mentioning a server with non-human content (not ERP) will get people complaining at you in OOC.

**Comment by maikilangiolo** [comment]
Cool on the surface but puzzling. Grabs become even better because they break the block (already grabs make melee users cry), toughness farm becomes even better (more HP and more block), large melee weapons become better (all swords are normal sized, axes and hammers and saws are max size).

I'd honestly either make the block last X seconds when pressed, but cannot be broken with a grab, with a cooldown, or just add passive parry/block chance. Can't we look into IS12 melee stances? They have a bunch of things including parry/block stances that seem already balanced, which would save us the trouble.

Overall, block is interesting, but I'm skeptical of how this PR implements it. No idea why anyone would bother to get long lasting slowdown, waiting to be hit, when the enemy gains am easier chance to run away, grab you, shoot you etc. It's probably only useful for people being ambushed with melee, rather than melee users.

**Comment by ErisMeleeEnjoyer** [comment]
> https://github.com/Citadel-Station-13/Citadel-Station-13/blob/0bfb671f4e4239818120b69768546849dfc990ba/code/modules/mob/living/living_active_block.dm

This vile thing uses stamina combat. Multiple people told both me and SPCR to NOT bring stamina combat to Eris. I share their opinion.

> i use swapping hands on MMB relatively often

how. Anyway, hotkey is F now, MMB is not affected.

> Also if we add blocking without any visual indicators it's kind of shit.

Visual indicator has been added.

> TGH is a great stat to have already, this basically makes TGH the best stat to max since scaling means at +100 TGH you don't go into paincrit AND can barehand block pretty much any melee attack.

It takes a lot of time(especially with gym's double insight capacity) to level up your stats. If someone goes from 0 to 100 TGH, well, congratulations to them - it takes an hour, or maybe even two, to do that. In comparison, an instagib cornbomb takes 30 minutes to make.

> Grabs become even better because they break the block (already grabs make melee users cry), toughness farm becomes even better (more HP and more block), large melee weapons become better (all swords are normal sized, axes and hammers and saws are max size).

I am going to nerf grabs by making counter-grabs work with items as well. I'm still thinking on how better to design that, but in general, grabs are going to be nerfed.

> I'd honestly either make the block last X seconds when pressed, but cannot be broken with a grab, with a cooldown, or just add passive parry/block chance.

Lag prevents short-window blocks from being not viable or good, it prevents them from being usable at all. This is why it has unlimited time - to prevent lag from turning blocking into an irksome mess. And because of unlimited use window, it slows down, so that people won't run around with their guard always up, evading ambushes.

> Can't we look into IS12 melee stances? They have a bunch of things including parry/block stances that seem already balanced, which would save us the trouble.

Too much clutter, honestly. We don't focus on melee as much as they do, nor do we have their med/combat system.

> No idea why anyone would bother to get long lasting slowdown, waiting to be hit, when the enemy gains am easier chance to run away, grab you, shoot you etc. It's probably only useful for people being ambushed with melee, rather than melee users.

You can turn blocking off at any given moment. If you hold on blocking for too long, you get countered by the opponent either running away, shooting you(how did they end up in a melee brawl then anyway?), or grabbing you. Yet, if you BAIT someone into grabbing you by walking around with blocking on, but then quickly turn blocking off right before someone grabs you, you can counter their grab, and turn the tables.

> This whole PR stinks of "I did not think through the consequences of this feature and how it would effect the game balance."

Yes, _melee_ game balance(we don't have any game balance in the first place since most of the good stuff like fireaxes/deckards is available roundstart) is affected here, for a videogame that focuses on _ranged_ weaponry. Oh, someone is blocking? Run away and make them chase you. Better yet, use a gun, since everyone has a gun and barehand blocking, unlike shields, cannot block projectiles.

**Comment by TheShown911** [comment]
Too much text, not reviewing, I'll leave it to people to argue forever so important stuff can be done without constant arguing(like closing all the new gun PRs)

**Comment by Alliostra** [comment]
Blocking just makes using melee weapons even less worthwhile compared to just using guns. Of course you could buff melee weapon damage to compensate, but then you'd just get more complaints by people who gotkilled in 3 hits by a surprise gamer chainsaw because they couldn't react fast enough, or because didn't want to spend all round grinding for stats.

**Comment by Pink-Chink** [comment]
Will test this again with sledgehammers after the weekend.

**Comment by Pink-Chink** [comment]
Been playing around with it a bit. Pretty solid as far as the gameplay goes. Any reason why there's nothing for blocking non-player mob attacks? It felt weird that I couldn't block them. Also, if I'm blocking and my attack is blocked by a shield (shield user wasn't blocking), my guard will remain up. I think guards should drop on attack no matter what.

Once those issues are addressed, I think this is good from a code perspective. I doubt there is anything more to learn from testing.

File: code/modules/mob/mob_movement.dm
Comment (Pink-Chink) [inline code/modules/mob/mob_movement.dm:118]:
```suggestion
	if(!mob.stat && isturf(mob.loc) && !mob.restrained())
```

File: code/modules/mob/living/carbon/human/human.dm
Comment (Pink-Chink) [inline code/modules/mob/living/carbon/human/human.dm:1786]:
```suggestion
	if(HUDneed.Find("block"))
```

File: code/modules/mob/living/carbon/human/human.dm
Comment (Pink-Chink) [inline code/modules/mob/living/carbon/human/human.dm:1775]:
```suggestion
	if(HUDneed.Find("block"))
```

File: code/_onclick/hud/screen_objects/base_screen_objects.dm
Comment (Pink-Chink) [inline code/_onclick/hud/screen_objects/base_screen_objects.dm:959]:
```suggestion
		if(H.blocking)
```

**Comment by Humonitarian** [comment]
resident balanceman here

Make disarms disrupt blocking, make grabs get completely negated instead.
Bonus points if the slowdown lingers half a second after you are disrupted by a disarm, allowing an easier follow-up grab.

If you want to be even more thorough, make disarm disrupt with an extended linger (maybe even a full second), whereas grab just breaks blocking. That way disarms are better, but mindlessly spamming grabs works too.

This will increase the depth to grab-fu, and make people think a little more about their moves. As well as, improves disarm utility, and nerfs grabs when compared to lethals.

Additionally, add a visually shown cooldown on blocking.
That is, while your blocking is on cooldown, the button should become red, indicating you cannot use that ability yet.
Such a system could later be implemented for jumping too, as well as improving sliding mechanics. As Reere intended.
(The only real issue with cooldown based moves is bad indication to the player - they have to know when their ability becomes usable, so they don't accidentally mess up trying to use it while on cooldown)

**Review by Pink-Chink — APPROVED** [review]
Blocked

**Pull request merged**
